### PR TITLE
Add support save images

### PR DESCRIPTION
**Pull request by jimmyxian**
Add support save images

### CODE REVIEW

**Comment by jimmyxian** [comment]
ping @aluzzardi @vieux @abronan

File: test/integration/api/save.bats [hunk 1]
```diff
@@ -42,3 +42,31 @@ function teardown() {
...
 	rm -f $temp_file_name_o
 }
+
+@test "docker save muti-images" {
+	start_docker_with_busybox 1
+	start_docker 1
+	swarm_manage
+	
+	# tag busybox
+	run docker_swarm tag busybox testimage
```
Comment (aluzzardi) [inline test/integration/api/save.bats]: So, in order to make sure that `save` actually looks up at different machines I would probably do something like:

```
start_docker_with_busybox 2
docker -H ${HOSTS[0]} tag busybox test1
docker -H ${HOSTS[0]} tag busybox test2
docker_swarm save test1 test2
```

This makes sure `save` is looking at both hosts.

**Comment by aluzzardi** [comment]
@vieux You are way more familiar than me with what's that supposed to do, could you take a look?

What kind of things should be tested? Maybe a load afterwards on a third (clean host, without even `busybox`, just `start_docker`), run `images` on it and make sure both `test1` and `test2` are there?

File: test/integration/api/save.bats (hunk 1 above)
Comment (jimmyxian) [inline test/integration/api/save.bats]: Well, does not support multi images on multiple nodes. :)

File: api/handlers.go [hunk 2]
```diff
@@ -62,6 +62,59 @@ func getVersion(c *context, w http.ResponseWriter, r *http.Request) {
...
+	if err := r.ParseForm(); err != nil {
+		httpError(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	names := r.Form["names"]
+
+	// find an engine which has all images
+	// Engine.Addr : [found names]
+	dict := make(map[string][]string)
+	bFoundEngine := false
```
Comment (chanwit) [inline api/handlers.go]: I think we do not need this `bFoundEngine`.
Just check with `foundEngineAddr = ""` as Golang alreay inits it to an empty string.

File: api/handlers.go [hunk 3]
```diff
@@ -62,6 +62,59 @@ func getVersion(c *context, w http.ResponseWriter, r *http.Request) {
...
+	var foundEngineAddr string
+	for _, image := range c.cluster.Images() {
+		for _, name := range names {
+			if image.Match(name) {
+				// check if engine addr already exists
+				value, exists := dict[image.Engine.Addr]
+				if exists {
+					// check if name already exists
+					found := false
```
Comment (chanwit) [inline api/handlers.go]: Could you change this to something like `nameAlreadyExisted` instead of `found`?

File: api/handlers.go [hunk 4]
```diff
@@ -62,6 +62,59 @@ func getVersion(c *context, w http.ResponseWriter, r *http.Request) {
...
+func getImages(c *context, w http.ResponseWriter, r *http.Request) {
+	if err := r.ParseForm(); err != nil {
+		httpError(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	names := r.Form["names"]
+
+	// find an engine which has all images
+	// Engine.Addr : [found names]
+	dict := make(map[string][]string)
```
Comment (chanwit) [inline api/handlers.go]: Variable named `dict` is too general. I guess it's the `imageMap` ?

**Comment by aluzzardi** [comment]
Regarding my previous comment on saving images which are on different machines:

1/ If it's something we want to support, we should add a test
2/ If it's something we **don't** want to support, we should add a test anyway and verify that it fails properly

**Comment by vieux** [comment]
@jimmyxian @aluzzardi I say we don't support save on multi machine (at least for now) so we it would be nice to add a proper error and a test.

**Comment by vieux** [comment]
Also, can we then rmi busybox and testimage; load the tar file; and do a images again ?

File: api/handlers.go (hunk 4 above)
Comment (jimmyxian) [inline api/handlers.go]: OK

File: api/handlers.go (hunk 2 above)
Comment (jimmyxian) [inline api/handlers.go]: OK

**Comment by jimmyxian** [comment]
@aluzzardi  @vieux Yeah. I will update by the comments.

**Comment by jimmyxian** [comment]
ping @aluzzardi @vieux @chanwit Modified by the comments, check again, thks

**Pull request closed without merging**